content/blog/2019-10-23-tail-call-elimination.md

```diff
@@ -0,0 +1,239 @@
+labels in the program. This way, if we have a tail call, then we can jump to
+the beginning of the callee instead of creating a new stack frame. This is [how
+tail call elimination is implemented](https://www.cs.cornell.edu/courses/cs4120/2019sp/lectures/34functional/lec34-sp19.pdf?1556325712)
+in x86 assembly.
+
+We have to modify the grammar to support features like pushing onto the stack
+and defining functions, then update the interpreter. Thanks to the work done by
+Alexa and Greg for [Project 1](https://github.com/sampsyo/bril/pull/16), these
+changes were much easier for me to make.
+
+## Implementation
+
+### Modifying the Grammar
+The first step is to modify the grammar to support function declarations with
+arguments and return types, as well as support the various new value/effect
+operations. Functions look as follows:
+```
+int foo(x: int, b: bool) {
+  print x;
+  print b;
+  ...
+}
+```
+Functions must specify a return type, a name, and a potentially empty list of
+typed arguments.
+
+Here are the new effect operations and their semantics:
+- `push arg1 ... argn`: push arguments to the stack, which can be used by the
+first function to be `call`'d after this instruction.
+- `call foo`: starts executing instructions defined by the function `foo`
+
+Here are the new value operations and their semantics:
+- `retval`: Retrieve the return value of the previous function call, e.g.
+```
+...
+call foo
+r: int = retval;
+```
+Here, if `foo` returned 0, then `r` would have the value 0.
+
+### Extending the Interpreter
+The interpreter needed to be extended to implement the above semantics. To model
+a stack frame, I explicitly keep track of the program counter (i.e. which
+instruction is being interpreted), the name of the current function, and an
+environment. I made this explicit because it made jumping to other functions
+easier to implement. The interpreter simply tries to evaluate the frame at the
+top of the stack until the stack is empty.
+
+Arguments that are `push`ed stay on the current stack frame. Once a `call` is
+made, we create a new stack frame. Note that the names of the `push`ed arguments
+don't necessarily match those declared by the function, so we need to map the
+function's arguments to the values of the `push`ed arguments.
+
+To return a value, a special variable name is set in the environment in the
+previous stack frame so it can be retrieved by the caller using `retval`.
+
+### Extending the TypeScript Frontend
+For the majority of this, I referred to [Alexa and Greg's implementation](https://github.com/sampsyo/bril/pull/16).
+There were some differences however because I have separate instructions for
+passing arguments to a function. So, whenever a function call was found in the
+AST, the arguments needed to be converted to Bril instructions first, then those
+would be used in a `push`. Additionally, a `retval` would need to be created
+afterwards if in the AST, the function call was assigned to a variable.
```

Comment:
```suggestion
afterwards if, in the AST, the function call was assigned to a variable.
```
The *result* of the function call, for example? And it doesn't seem to matter whether or not this is assigned to a variable… for example, you'd want to get the return value of `f` if you did `g(f() + x)` or something, i.e., using the result for something other than an assignment.